tools/validators/instance_validator/validate/entity_instance.py

```diff
@@ -15,293 +15,395 @@
+        is_valid = False
+
+    return is_valid
+
+
+  def _ValidateLinks(self, universe, entity_instances):
+    """Uses information from the generated ontology universe to validate
+    the links key of an entity.
+
     Args:
-      entity_instances: dictionary containing all instances.
+      universe: ConfigUniverse generated from the ontology
+      entity_instances: dictionary containing all instances
+
     Returns:
-      Returns boolean for validity of entity.
+      Returns boolean for validity of links key, defaulting to True if the
+      key is not present.
     """
-    for req_key in self.required_keys:
-      if req_key not in self.entity.keys():
-        print('Missing required key:', req_key)
-        return False
 
-    if not self._ValidateType():
-      print('Invalid type')
-      return False
+    if self.links is None:
+      return True
 
-    if not self._ValidateLinks(entity_instances):
-      print('Invalid links')
-      return False
+    is_valid = True
 
-    if not self._ValidateTranslation():
-      print('Invalid translation')
-      return False
+    entity_type = universe.GetEntityType(self.namespace, self.type_name)
+    type_fields = entity_type.GetAllFields()
+    found_fields = set()
+
+    for link_inst in self.links:
+      if link_inst.source not in entity_instances.keys():
+        print('Invalid link source entity name: {0}'.format(link_inst.source))
+        is_valid = False
+        continue
+
+      src_entity = entity_instances.get(link_inst.source)
+      src_namespace = src_entity.namespace
+      src_type_name = src_entity.type_name
+      src_entity_type = universe.GetEntityType(src_namespace,
+                                                    src_type_name)
+
+      for source_field, target_field in link_inst.field_map.items():
+        # assumes that the namespace is '' for now
+        if not entity_type.HasField('/' + target_field):
+          print('Invalid link target field: ', target_field)
+          is_valid = False
+          continue
+
+        if not src_entity_type.HasField('/' + source_field):
+          print('Invalid link source field: ', source_field)
+          is_valid = False
+          continue
+
+        found_fields.add('/' + target_field)
 
-    if not self._ValidateConnections():
-      print('Invalid connections')
+        if not self._ValidateLinkUnits(universe, source_field, target_field):
+          is_valid = False
+          continue
+
+        if not self._ValidateLinkStates(universe, source_field, target_field):
+          is_valid = False
+          continue
+
+    for field_name, field in type_fields.items():
+      if not field.optional and field_name not in found_fields:
+        print('Required field {0} is missing from links'.format(field_name))
+        is_valid = False
+
+    return is_valid
+
+
+  def _ValidateLinkUnits(self, universe, source_field, target_field):
+    """Validates that units match between linked source and target fields."""
```

Comment:
I am not sure I follow this one, the units are on the translation only? they dont belong to links:
```VLV-141:
  id: CDM/2knv7XuT1DPAxr33jYj$0f
  links:
    FCU-92:
      chilled_water_flowrate_sensor: chilled_water_flowrate_sensor
      chilled_water_valve_percentage_command: chilled_water_valve_percentage_command
      chilled_water_valve_percentage_sensor: chilled_water_valve_percentage_sensor
      discharge_air_temperature_sensor: discharge_air_temperature_sensor
      return_air_temperature_sensor: return_air_temperature_sensor```

Reply:
maybe rename to ValidateUnits?

Reply:
This is validating that the source and target fields of the link have compatible units, so this is a distinct check from the translation validation. It's a helper method for ValidateLinks below.

I renamed it to ValidateLinkUnitsMatch - is that clearer?